File: kf.m
classdef kf < handle
    %KF A generic Kalman Filter class
    % This class implements all the basic Kalman Filter equations, but uses
    % State Block and Measurement Block objects to define all the matrices
    % and state propagation functions.
    %
    % Author: Walter Livingston

    properties
        state_block         (1,1)           % State Block Object
        measurement_block   (1,1)           % Measurement Block Object
        covariance_block    (1,1)           % Covariance Block Object
        x                   (:,1) double    % State Vector
        X                   (:,1) double    % Nominal State Vector
        P                   (:,:) double    % State Covariance Matrix
        z                   (:,1) double    % Innovation Vector
        mode                (1,1) string    % Kalman Filter Mode
    end
    
    methods
        function obj = kf(state_block, measurement_block, covariance_block, options)
            arguments
                state_block         (1,1) stateblock
                measurement_block   (1,1) measurementblock
                covariance_block    (1,1) covarianceblock
                options.xi          (:,1) double = zeros(state_block.num_states,1);
                options.X_i         (:,1) double = zeros(state_block.num_states,1);
                options.dt          (1,1) double = 1
                options.mode        (1,1) string = 'kf'
            end

            obj.state_block = state_block;
            obj.measurement_block = measurement_block;
            obj.covariance_block = covariance_block;
            obj.mode = options.mode;
            obj.x = options.xi;

            if strcmp(obj.mode,"error")
                obj.X = options.X_i;

                sAux = obj.state_block.aux;
                sAux.X = obj.X;

                mAux = obj.measurement_block.aux;
                mAux.X = obj.X;
                
                obj.state_block.applyError(obj.x, obj.X, options.dt);
                obj.measurement_block.applyError(obj.x, obj.X);

                obj.state_block = obj.state_block.processAuxData(sAux);
                obj.measurement_block = obj.measurement_block.processAuxData(mAux);
            end
            
            obj.P = covariance_block.calcProcessCovarianceMatrix(obj, options.dt);

            switch obj.mode
                case 'kf'
                    obj.state_block.F = obj.state_block.updateStateTransitionMatrix(options.xi);
                    obj.measurement_block.H = ...
                        obj.measurement_block.updateObservationMatrix(options.xi, ...
                            0);
                case 'extended'

            end
        end
        
        function [obj] = process(obj, dt, customStateBlock, customInnBlock)
        %PROCESS Implements the Time Update of a Kalman Filter
        % This function runs the time update of a Kalman filter. Using the
        % stateblock passed into the constructor, this function creates the
        % state transition and process covariance matrices and propagates
        % the state estimates and state covariance matrix.
            if exist('customStateBlock', 'var')
                sBlock = customStateBlock;
                obj.state_block = customStateBlock;
            else
                sBlock = obj.state_block;
            end

            if exist('customInnBlock', 'var')
                cBlock = customInnBlock;
            else
                cBlock = obj.covariance_block;
            end

            [obj.x, Phi] = sBlock.propagate(obj.x, dt, ...
                (strcmp(obj.mode, 'extended')) || (strcmp(obj.mode, 'error')));
            Qd = cBlock.calcProcessCovarianceMatrix(obj, dt, sBlock);
            obj.P = Phi*obj.P*Phi' + Qd;
            
            if strcmp(obj.mode,'error')
                obj.X = sBlock.applyError(obj.x, obj.X, dt);
                aux = sBlock.aux;
                aux.X = obj.X;
                obj.state_block = sBlock.processAuxData(aux);
            end
        end

        function [obj] = update(obj, y, customMeasBlock, customInnBlock)
            if exist('customMeasBlock', 'var')
                mBlock = customMeasBlock;
                aux = mBlock.aux;
                aux.X = obj.X;
                aux.y = y;
                obj.measurement_block = customMeasBlock.processAuxData(aux);
            else
                mBlock = obj.measurement_block;
            end

            if exist('customInnBlock', 'var')
                cBlock = customInnBlock;
            else
                cBlock = obj.covariance_block;
            end

            [yhat, H] = mBlock.update(obj.x, y, ...
                (strcmp(obj.mode, 'extended')) || (strcmp(obj.mode, 'error')));
            if any(isnan(obj.P)); error(''); end
            R = cBlock.calcMeasurementCovarianceMatrix(obj, mBlock);
            oldz = obj.z;
            obj.z = (y - yhat);
            [S,reject] = cBlock.calcInnovationCovarianceMatrix(obj);
            if ~reject
                L = obj.P*H'/S;
    
                obj.x = obj.x + L*obj.z;
                obj.P = (eye(obj.state_block.num_states) - L*H)*obj.P*(eye(obj.state_block.num_states) - L*H)' + L*R*L';
    
                if strcmp(obj.mode, 'error')
                    [obj.x, obj.X] = mBlock.applyError(obj.x, obj.X);
                    aux = mBlock.aux;
                    aux.X = obj.X;
                    aux.y = y;
                    obj.measurement_block = mBlock.processAuxData(aux);
                end
            else
                obj.z = oldz;
            end
        end
    end
end

File: modular_kf/kf.py
from __future__ import annotations
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from modular_kf import CovarianceBlock

import numpy as np
from modular_kf import StateBlock, MeasurementBlock
from modular_kf.helpers import *

class KalmanFilter:
    state_block: StateBlock
    measurement_block: MeasurementBlock
    covariance_block: CovarianceBlock
    x: vec = vec0
    X: vec = vec0
    z: vec = vec0
    P: mat = mat0
    mode: str = ""

    def __init__(self, state_block, measurement_block, covariance_block, **options):
        self.state_block: StateBlock = state_block
        self.measurement_block: MeasurementBlock = measurement_block
        self.covariance_block = covariance_block

        self.x = options.get('x_i', np.zeros((state_block.num_states, 1)))
        self.dt = options.get('dt', 0)
        self.mode = options.get('mode', "linear")

        self.P = self.covariance_block.calcProcessCovarianceMatrix(self.dt)

        match self.mode:
            case "linear":
                self.state_block.F = self.state_block.updateStateTransitionMatrix(self.x)
                self.measurement_block.H = self.measurement_block.updateObservationMatrix(self.x, vec0)
            case "extended":
                pass
            case "error":
                self.X = options.get('X_i', np.zeros((state_block.num_states, 1)))

                sAux = self.state_block.aux
                sAux.X = self.X

                mAux = self.measurement_block.aux
                mAux.X = self.X

                self.state_block = self.state_block.processAuxData(sAux)
                self.measurement_block = self.measurement_block.processAuxData(mAux)

                self.state_block.applyError(self.x, self.X, self.dt)
                self.measurement_block.applyError(self.x, self.X)
            case _:
                raise ValueError("This type of Kalman Filter is undefined!")
            
    def process(self, dt: float, customStateBlock: StateBlock | None = None, customCovBlock: CovarianceBlock | None = None):
        if customStateBlock is not None:
            self.state_block = customStateBlock

        if customCovBlock is not None:
            self.covariance_block = customCovBlock

        [self.x, Phi] = self.state_block.propagate(self.x, dt, self.mode == 'extended' or self.mode == 'error')
        Qd = self.covariance_block.calcProcessCovariance(self, dt, self.state_block)
        P = Phi*self.P*Phi.T + Qd
        self.P = np.matrix(P)

        if self.mode == 'error':
            ewn = self.state_block.applyError(self.x, self.X, dt)
            self.X = ewn.E
            self.x = ewn.E

            aux = self.state_block.aux
            aux.X = self.X
            self.state_block.processAuxData(aux)

    def update(self, y: vec, customMeasBlock: MeasurementBlock | None = None, customCovBlock: CovarianceBlock | None = None):
        if customMeasBlock is not None:
            self.measurement_block = customMeasBlock

        if customCovBlock is not None:
            self.covariance_block = customCovBlock

        [yhat, H] = self.measurement_block.update(self.x, y, self.mode == 'extended' or self.mode == 'error')
        R = self.covariance_block.calcMeasurementCovariance(self, self.measurement_block)
        oldz = self.z
        self.z = (y - yhat)
        [S,reject] = self.covariance_block.calcInnovationCovariance(self)

        if not reject:
            L = self.P*H.T/S

            self.x = self.x + L*self.z
            P = (np.eye(self.state_block.num_states) - L*H)*self.P*(np.eye(self.state_block.num_states) - L@H).T + L*R*L.T
            self.P = np.matrix(P)

            if self.mode == 'error':
                [self.x, self.X] = self.measurement_block.applyError(self.x, self.X)
                aux = self.measurement_block.aux
                aux.X = self.X
                aux.y = y
                self.measurement_block.processAuxData(aux)
        else:
            self.z = oldz


